metaflow/plugins/kfp/kfp.py

```diff
@@ -1,86 +1,293 @@
+
+        command_template_map[cur_step] = build_cmd_template(cur_step, cur_task_id, cur_input_path)
+
+        for step in cur_node.out_funcs:
+            if step not in seen_steps:
+                steps_deque.append(step)
+                seen_steps.add(step)
+
+    return command_template_map
+
+
+def create_flow_from_graph(flowgraph, flow_code_url=DEFAULT_FLOW_CODE_URL):
+
+    graph = flowgraph
     code_url = flow_code_url
-    print("\nCreating the pipeline definition needed to run the flow on KFP...\n")
-    print("\nCode URL of the flow to be converted to KFP: {0}\n".format(flow_code_url))
+
+    command_template_map = create_command_templates_from_graph(graph)
 
     @dsl.pipeline(
         name='MF on KFP Pipeline',
-        description='Pipeline defining KFP equivalent of the Metaflow flow'
+        description='Pipeline defining KFP equivalent of the Metaflow flow. Currently supports linear flows and flows '
+                    'with branch and join nodes'
     )
     def kfp_pipeline_from_flow():
-        """
-        This function converts the flow steps to kfp container op equivalents
-        by invoking `step_container_op` for every step in the flow and handling the order of steps.
-        """
+        S3_BUCKET = V1EnvVar(name="S3_BUCKET", value=S3_BUCKET_VALUE)
+        S3_AWS_ARN = V1EnvVar(name="S3_AWS_ARN", value=S3_AWS_ARN_VALUE)
+        pre_start_op = (pre_start_container_op())(code_url).add_env_variable(S3_BUCKET).add_env_variable(S3_AWS_ARN)
+        outputs = pre_start_op.outputs
+
+        container_op_map = {}
+        ds_root = outputs['ds_root']
+        run_id = outputs['run_id']
+
+        # Initial setup
+        pre_start_op.set_display_name('InitialSetup')
+        container_op_map['start'] = (step_container_op())(
+                                        command_template_map['start'],
+                                        'start',
+                                        code_url,
+                                        ds_root,
+                                        run_id
+                                        ).add_env_variable(S3_BUCKET).add_env_variable(S3_AWS_ARN)
+        container_op_map['start'].set_display_name('start')
+        container_op_map['start'].after(pre_start_op)
 
-        # Store the list of steps in reverse order
-        step_container_ops = [step_container_op(step, code_url) for step in reversed(steps)]
+        for step, cmd in command_template_map.items():
+            if step != 'start':
+                container_op_map[step] = (step_container_op())(
+                                            command_template_map[step],
+                                            step,
+                                            code_url,
+                                            ds_root,
+                                            run_id).add_env_variable(S3_BUCKET).add_env_variable(S3_AWS_ARN)
```

Comment:
Can apply the environment variables to all pipeline ops using a KFP transform, 
see example: https://github.com/kubeflow/pipelines/blob/master/samples/core/pipeline_transformers/pipeline_transformers.py#L39

Reply:
Done. Thank you - this is cleaner and more readable